core/src/main/java/io/grpc/internal/DelayedClientTransport.java

```diff
@@ -257,7 +253,7 @@ void setError(Status reason) {
           }
         }
       }
-      super.setError(reason);
+      super.cancel(reason);
```

Comment:
I should have noticed before, but we should cancel the stream before reporting transport terminated (so call super first).